File: Data Structures & Algorithms/longest-increasing-subsequence/submission-1.cpp
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        int n = nums.size();
        vector<int> dp(n+1, -1);

        int ans = 1;
        for(int i = 0;i<n;i++) {
            ans = max(ans, dfs(nums, dp, i));
        }
        return ans;
    }

    int dfs(vector<int>& nums, vector<int>& dp, int i) {
        if(dp[i] != -1) return dp[i];

        int ans = 1;

        for(int j = i+1;j<nums.size();j++) {
            if(nums[i] < nums[j])
                ans = max(ans,1+ dfs(nums, dp, j));
        }

        return dp[i] = ans;
    }
};

File: Data Structures & Algorithms/permutations/submission-1.java
class Solution {
    public List<List<Integer>> permute(int[] nums) {
        int n = nums.length;
        List<List<Integer>> ans = new ArrayList<>();
        
        dfs(nums, n - 1, ans);
        return ans;
    }

    private void dfs(int[] nums, int i, List<List<Integer>> ans) {
        if(i == 0) {
            List<Integer> currentPerm = new ArrayList<>();
            for(int num : nums) {
                currentPerm.add(num);
            }
            ans.add(currentPerm);
            return;
        }

        if(i < 0) return;

        for (int j = i; j >= 0; j--) {
            swap(nums, i, j);
            dfs(nums, i - 1, ans);
            swap(nums, i, j);
        }
    }

    private void swap(int[] nums, int a, int b) {
        int temp = nums[a];
        nums[a] = nums[b];
        nums[b] = temp;
    }
}